For case exports show case property group

mkangia opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
For case exports,
for any export column that is generated for a case property, show corresponding case property group from Data dictionary as a tag, if present.

This change is only applicable for projects with access to data dictionary & is skipped for bulk case exports where we don't show columns to the user on UI.

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi.atlassian.net/browse/SC-3452

A small change to how this was planned originally.
Initially the requirement was to "group" export columns just like properties on data dictionary.
However, case exports are customizable when it comes to ordering i.e a user is allowed to move columns around. Considering that, explicit grouping would be conflicting since user would want to order things differently.
Though, in order to still provide the user with the information about the groups, and then possibly group columns accordingly, we can still show the group name, this way, they can still choose to group columns together or leave them separate.

--------

Example:
(Notice the "Details" / "Others" tags added which are from data dictionary)

![image](https://github.com/user-attachments/assets/44d9806d-742f-47c8-9613-b9447c3a04d4)


## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Tested locally.
The change is limited to domains with access to data dictionary and is further limited to non-bulk case exports.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

Added tests for the changes.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
Will take it through QA.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

mkangia committed:
store case_property_group_name on ScalarItem for case properties

mkangia committed:
show case_property_group_name as a label when available

mkangia committed:
skip adding case prop group for bulk export

mkangia committed:
dedup schema generation

mkangia committed:
fix tests

Charl1996 commented on corehq/apps/data_dictionary/util.py:
```diff
@@ -417,3 +417,15 @@ def get_used_props_by_case_type(domain):
+
+def get_case_property_group_name_for_properties(domain, case_type_name):
+    case_group_name_for_property = {}
+    for case_property_name, group_name in (
```
nit: what do you think about pulling out the query construction which will be looped over into a separate variable, then using that variable in the for-loop statement? 

```suggestion
case_property_group_query = CaseProperty.objects.filter(
     case_type__name=case_type_name, case_type__domain=domain
 ).select_related('case_type').select_related('group').values_list('name', 'group__name')
 
 for case_property_name, group_name in case_property_group_query:
```

I bring this up because when I started reading the for-loop I was stuck trying to figure out what the query does whilst having to remember that I'm reading this in the context of a for-loop. It would be nice if I could mentally separate these two things.

Up to you.

mkangia replied:
Totally valid, thanks for sharing the context as well. :+1: 

Include the change that you suggested.

mkangia committed:
extract query

mkangia committed:
move property group to its own column

mkangia committed:
add help text to header

mkangia commented:
Hey @Charl1996 

Thanks for review on this so far. This PR is now ready for review since we have a soft agreement on the approach. If we go through with this approach considering agreement from SaaS, I don't expect much to change so good to give it a full code review.

mkangia committed:
rename to generic export name

mkangia committed:
better styling

zandre-eng commented on corehq/apps/export/models/new.py:
```diff
@@ -2442,12 +2448,17 @@ def _generate_schema_from_case_property_mapping(cls, case_property_mapping, pare
                     path=[PathNode(name=prop)],
                     label=prop,
                     last_occurrences={app_id: app_version},
+                    case_property_group_name=case_group_name_for_property.get(prop, None)
```
Nit: The default of `None` can be removed since this is already the default return val for `get()`.

zandre-eng commented on corehq/apps/export/templates/export/partials/new_customize_export_templates.html:
```diff
@@ -214,6 +214,11 @@ <h4>{{ terminology.choose_fields_label }}</h4>
               {% trans "Property Group" %}
+              <span
+                data-bind="makeHqHelp: {
+                  description: '{% trans "From case property groups on Data Dictionary" %}'
```
Nit: Small grammar change, "From case property groups **in** Data Dictionary".

zandre-eng reviewed (approved):
Looks good, just some minor non-blocking feedback.

esoergel commented on corehq/apps/data_dictionary/util.py:
```diff
@@ -417,3 +417,15 @@ def get_used_props_by_case_type(domain):
+    case_group_name_for_property = {}
+    case_property_group_query = CaseProperty.objects.filter(
+        case_type__name=case_type_name, case_type__domain=domain
+    ).select_related('case_type').select_related('group').values_list('name', 'group__name')
```
What do the `select_related` calls do here if we're only returning two columns?

mkangia replied:
You are right. I didn't realize django would already handle that for the query and values_list.

esoergel commented on corehq/apps/data_dictionary/util.py:
```diff
@@ -417,3 +417,15 @@ def get_used_props_by_case_type(domain):
+
+
+def get_case_property_group_name_for_properties(domain, case_type_name):
+    case_group_name_for_property = {}
+    case_property_group_query = CaseProperty.objects.filter(
+        case_type__name=case_type_name, case_type__domain=domain
+    ).select_related('case_type').select_related('group').values_list('name', 'group__name')
+
+    for case_property_name, group_name in case_property_group_query:
+        case_group_name_for_property[case_property_name] = group_name
+
+    return case_group_name_for_property
```
Nit: I think this is equivalent:
```suggestion
    return dict(CaseProperty.objects.filter(
        case_type__name=case_type_name, case_type__domain=domain
    ).values_list('name', 'group__name'))
```

mkangia replied:
Beautiful. Thank You.

esoergel reviewed (approved):
> Initially the requirement was to "group" export columns just like properties on data dictionary.
> However, case exports are customizable when it comes to ordering i.e a user is allowed to move columns around. Considering that, explicit grouping would be conflicting since user would want to order things differently.

It definitely makes sense to allow users to reorder (and delete) columns in the export configuration.  However, with new exports, the only user-provided ordering we have is the one from the data dictionary.  Does it already use the data-dictionary ordering for new exports, so properties in the same group are next to each other and in the order they appear on the DD page?

mkangia committed:
Simplify query; remove redundant select related

Co-authored-by: Ethan Soergel <esoergel@users.noreply.github.com>

mkangia committed:
minor code improvements

mkangia commented:
Hey @esoergel 

> However, with new exports, the only user-provided ordering we have is the one from the data dictionary. Does it already use the data-dictionary ordering for new exports, so properties in the same group are next to each other and in the order they appear on the DD page?

Ordering was added [here](https://github.com/dimagi/commcare-hq/commit/50c537217d2c08be51cfc86312426342f772f742) for new case exports, after which like you rightly said, same group properties are next to each other for new case exports.
The default [fetching of properties](https://github.com/dimagi/commcare-hq/blob/055a014e090dde959c91a47ea13ab7704fba9f1f/corehq/apps/app_manager/app_schemas/case_properties.py#L404C25-L404C57), does not have any ordering of its own.

mkangia committed:
skip deprecated props and groups

mkangia commented:
Hey @zandre-eng @esoergel 

Made a small change to exclude deprecated groups and properties. Requesting a review on the new commit(s) added.

mkangia committed:
revert: move property group to its own column

mkangia committed:
merge master; resolve conflicts

mkangia committed:
change color for group label

mkangia committed:
reset group names on current schema

mkangia commented:
Hey @zandre-eng @esoergel 
Requesting for another pass at review. There have been couple of followups.

Summary of changes since last review:
1. From [feedback from other teams](https://docs.google.com/document/d/1h6Fxqp2fcQ7rAM7C44GFLAhCJk4QMnYvhrNIGslRhRA/edit?disco=AAABQRn6u_0) we decided to keep group names as tags in the Property column itself instead of having them as a separate column (Screenshot updated in the PR description)
2. a slight [change](https://docs.google.com/document/d/1h6Fxqp2fcQ7rAM7C44GFLAhCJk4QMnYvhrNIGslRhRA/edit?disco=AAABOURfeqQ) to the color of this tag that holds the property group name so its differentiable from usual tag (Screenshot updated in the PR description)
3. resetting group names before setting them to clear them out for deleted/deprecated properties/groups. This resolves this [bug](https://dimagi.atlassian.net/browse/SC-3746) reported.

mkangia committed:
dedup: extract logic

mkangia committed:
set static method

zandre-eng commented on corehq/apps/export/models/new.py:
```diff
@@ -2434,6 +2440,14 @@ def _reorder_case_properties_from_data_dictionary(cls, domain, case_type, case_p
 
+    @classmethod
+    def _reset_case_property_group_name(cls, schema):
+        for group_schema in schema.group_schemas:
```
Nit: There's a lot of nesting happening in this loop. One level of nesting can be removed by rearranging the code as:
```
for group_schema in schema.group_schemas:
    if group_schema.path != MAIN_TABLE:
        continue
    for item in group_schema.items:
        if isinstance(item, ScalarItem):
            item.case_property_group_name = ''
```

This is completely personal preference however, and happy for the code to stay as is.

zandre-eng reviewed (approved):
Changes look good. Thanks for all the iteration on this Manish.

mkangia commented:
Thanks @Charl1996
